common/formatting/align.cc

```diff
@@ -117,6 +130,133 @@ std::ostream& operator<<(std::ostream& stream, const AlignmentCell& cell) {
+        const VectorTree<ValueType>&)>& get_cell_label) {
+  if (root.Children().empty()) return;
+
+  static constexpr absl::string_view kCellSeparator = "|";
+
+  struct Cell {
+    std::string text = "";
+    char filler = ' ';
+    std::size_t width = 0;
+  };
+  VectorTree<Cell> text_tree;
+
+  const std::size_t depth =
+      CreateTextNodes<ValueType, Cell>(root, &text_tree, get_cell_label);
+
+  // Adjust cells width to fit all their children
+  for (auto& node : VectorTreePostOrderTraversal(text_tree)) {
+    // Include separator width in cell width
+    node.Value().width += kCellSeparator.size();
+    if (node.is_leaf()) continue;
+    std::size_t children_width =
+        std::accumulate(node.Children().begin(), node.Children().end(), 0,
+                        [](std::size_t width, const VectorTree<Cell>& child) {
+                          return width + child.Value().width;
+                        });
+    if (node.Value().width < children_width) {
+      node.Value().width = children_width;
+    }
+  }
+  // Adjust cells width to fill their parents
+  for (auto& node : VectorTreePreOrderTraversal(text_tree)) {
+    if (node.is_leaf()) continue;
+    std::size_t children_width =
+        std::accumulate(node.Children().begin(), node.Children().end(), 0,
+                        [](std::size_t width, const VectorTree<Cell>& child) {
+                          return width + child.Value().width;
+                        });
+    if (node.Value().width > children_width) {
+      auto extra_width = node.Value().width - children_width;
+      for (auto& child : node.Children()) {
+        const auto added_child_width =
+            extra_width * child.Value().width / children_width;
+        extra_width -= added_child_width;
+        children_width -= child.Value().width;
+        child.Value().width += added_child_width;
+      }
+    }
+  }
+
+  std::vector<std::string> lines(depth);
+  auto range = VectorTreePreOrderTraversal(text_tree);
+  auto level_offset = text_tree.NumAncestors() + 1;
+  for (auto& node : make_range(range.begin() + 1, range.end())) {
+    auto& cell = node.Value();
+    const std::size_t level = node.NumAncestors() - level_offset;
+    if (level > 0 && node.IsFirstChild()) {
+      const int padding_len = lines[level - 1].size() - lines[level].size() -
+                              node.Parent()->Value().width;
+      if (padding_len > 0) {
+        if (lines[level].empty())
+          lines[level].append(std::string(padding_len, ' '));
+        else if (padding_len > int(kCellSeparator.size()))
+          lines[level].append(absl::StrCat(
+              kCellSeparator,
+              std::string(padding_len - kCellSeparator.size(), ' ')));
+      }
+    }
+
+    const std::vector<absl::string_view> parts =
+        absl::StrSplit(cell.text, '\t');
+    CHECK_LE(parts.size(), 3);
+
+    const auto width = cell.width - kCellSeparator.size();
+
+    if (parts.size() == 1) {
```

Comment:
Would `switch (parts.size()) / case ...` maybe more readable (I leave it up to you to play with it and decide).

Reply:
Fixed in [parent PR](https://github.com/chipsalliance/verible/pull/865)